contracts/CreditorProxy.sol

```diff
@@ -0,0 +1,409 @@
+/*
+
+  Copyright 2017 Dharma Labs Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity 0.4.18;
+
+import "./DebtKernel.sol";
+import "./DebtToken.sol";
+import "./TermsContract.sol";
+import "./TokenTransferProxy.sol";
+import "zeppelin-solidity/contracts/lifecycle/Pausable.sol";
+import "zeppelin-solidity/contracts/math/SafeMath.sol";
+import "zeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+
+
+/**
+ * The CreditorProxy is a thin wrapper around the DebtKernel
+ * It implements creditor-driven loans as specified by DIP-1
+ *
+ * Authors: Bo Henderson <bohendo> & Shivani Gupta <shivgupt>
+ */
+contract CreditorProxy is Pausable {
+    using SafeMath for uint;
+
+    enum Errors {
+        CREDIT_ORDER_CANCELLED,
+        CREDIT_ORDER_NON_CONSENSUAL,
+        CREDITOR_BALANCE_OR_ALLOWANCE_INSUFFICIENT
+    }
+
+    DebtToken public debtToken;
+    DebtKernel public debtKernel;
+
+    // solhint-disable-next-line var-name-mixedcase
+    address public TOKEN_TRANSFER_PROXY;
+    bytes32 constant public NULL_ISSUANCE_HASH = bytes32(0);
+    uint16 constant public EXTERNAL_QUERY_GAS_LIMIT = 8000;
+
+    mapping (bytes32 => bool) public creditOrderCancelled;
+
+    event LogCreditOrderCancelled(
+        address indexed _creditor,
+        bytes32 indexed _creditorCommitmentHash
+    );
+
+    event LogCreditOrderFilled(
+        address indexed _creditor,
+        bytes32 indexed _creditorCommitmentHash,
+        bytes32 indexed _agreementId
+    );
+
+    event LogError(
+        uint8 indexed _errorId,
+        address indexed _creditor,
+        bytes32 indexed _creditorCommitmentHash
+    );
+
+    struct CreditorCommitment {
+        address creditor;
+        address repaymentVersion;
+        uint creditorFee;
+        address underwriter;
+        uint underwriterRiskRating;
+        address termsContract;
+        bytes32 termsContractParameters;
+        uint commitmentExpirationTimestampInSec;
+        uint salt;
+        bytes32 hash;
+    }
+
+    function CreditorProxy(address _tokenTransferProxy)
+        public
+    {
+        TOKEN_TRANSFER_PROXY = _tokenTransferProxy;
+    }
+
+    /**
+     * Allows contract owner to set the currently used debt token contract.
+     * Function exists to maximize upgradeability of individual modules
+     * in the entire system.
+     */
+    function setDebtToken(address debtTokenAddress)
```

Comment:
We have a contract registry (https://github.com/dharmaprotocol/charta/blob/master/contracts/ContractRegistry.sol) which contains a reference to core contracts such as the DebtKernel and DebtToken. That's the pattern we've been using for upgradeability in the other contracts, so it warrants some discussion on why we would deviate from that pattern here.

Reply:
+1

Reply:
Ahh, can't believe we overlooked the contract registry, thanks for pointing that out.

We were following patterns we saw in the DebtKernel though, now I'm curious why it has a dedicated [`setDebtToken`](https://github.com/dharmaprotocol/charta/blob/master/contracts/DebtKernel.sol#L143) method rather than using the contract registry as well..

Reply:
I'm assuming this would also apply to the token transfer proxy. We'll get the contract registry in the constructor and then use that to get all the other contract addresses as needed like the example terms contracts do.

Similarly, I'm assuming we don't need a setter for the registry address since the terms contracts also omit it.